### ISSUE TITLE
feat: publicly export Deferred 

### ISSUE THREAD

**Issue by NickCrews** [ibis-project/ibis#7613]
### Is your feature request related to a problem?

I want to use this in type hints.

Before, I did `from ibis.expr.deferred import Deferred` but this recently broke, now I have to do `from ibis.common.deferred import Deferred`.

### Describe the solution you'd like

It would be great to export this stably. just `from ibis import Deferred` should work?

### What version of ibis are you running?

7.1.0

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by tswast** [comment]
I think this should be exposed in this API: `ibis.expr.types` as that's where I expect all the other expression-y types to be found.

**Comment by NickCrews** [comment]
That is fine with me as long as it is canonical and stable.

**Comment by kszucs** [comment]
Can you provide a couple of examples where you either use `Deferred` objects or check against its type? I have an assumption that you may rather need a function which transparently coerces various input objects - like `Deferred` - to expressions.

**Comment by tswast** [comment]
@kszucs FWIW, I encountered this here: https://github.com/googleapis/python-bigquery-dataframes/pull/53 I needed to call `get_name()` on an object that might or might not be Deferred:

```python
        self._column_names = {
            (
                column.resolve(table)
                # TODO(https://github.com/ibis-project/ibis/issues/7613): use
                # public API to refer to Deferred type.
                if isinstance(column, ibis.common.deferred.Deferred)
                else column
            ).get_name(): column
            for column in self._columns
        }
```

**Comment by NickCrews** [comment]
A polymorphic function that can either take python builtins int or float, or ibis types. Perhaps I'm not doing this the best way, if you have a way of avoiding the Deferred then I'm all ears.

```python
@overload
def round_to(nums: NumericValue | Deferred, n: int) -> NumericValue:
    ...


@overload
def round_to(nums: int | float, n: int) -> int:
    ...


def round_to(nums: NumericValue | Deferred | int | float, n: int) -> IntegerValue:
    """Round to the nearest n, eg 5."""
    if isinstance(nums, (NumericValue, Deferred)):
        return (nums / n).round().cast(int) * n
    else:
        return round(nums / n) * n


def test_round_to_deferred():
    x = ibis.memtable({"inp": [1, 2, 4], "expected": [0, 0, 5]})
    x = x.mutate(result=round_to(_.inp, 5))
    df = x.to_pandas()
    assert df["result"].equals(df["expected"])
```

**Comment by cpcloud** [comment]
@kszucs Any thoughts here based on feedback from Tim and Nick?

**Comment by NickCrews** [comment]
I am also using it as a type annotation for a dataclass like [`condition: bool | Deferred | Callable[[Table], BooleanValue]` ](https://github.com/NickCrews/mismo/blob/22bf177f23e39100d8c85d96f9de27f134b7a828/mismo/compare/_comparison.py#L31)

**Comment by NickCrews** [comment]
Another use case I have:

I have a Table -> Table function. I have some legacy pandas code that operates on DFs, but I want to use this new Table function. So I wrote
a decorator that makes it so a function takes either a Table or DF as input, converts it to a memtable if needed, calls the function, and reverts the result back to a DF if called with a DF. In those type checks i also need to worry about checking for Deferreds.

I have a similar decorator for Columns/pd.Series.

**Comment by jcrist** [comment]
This was fixed in #9008. Closing.
